### ISSUE TITLE
How to insert an operator(such as StopGradient) to pre-trained model?

### ISSUE THREAD

**Issue by LMerCy** [facebookarchive/caffe2#1744]
Hello 

I try to re-train the last fc-layer in a pre-trained Alex-net with predict_net.pb and exec_net.pb. But i don't know how to only train the last fc-layer with StopGradient.

 [Here ](https://nbviewer.jupyter.org/gist/kyamagu/6cff70840c10ca374e069a3a7eb00cb4)provides a method to train the layer you want  with method "create_param". This method is a little complex to me , and i don't know whether it works.

   I want to know is there any method to add StopGradient to a a loaded model.net? By this i can easily make sure stopgradient works by generating the graph of model.net. I tried to use model.StopGradient("fc7","fc7"), but it creates a new branch in the net, like this:

![image](https://user-images.githubusercontent.com/13573827/34857442-c015939e-f785-11e7-96e4-94be7a825a25.png)

i doubt whether it can stop backward from fc8 to fc7?

**Comment by LMerCy** [comment]
i have get some idea from issue#1017, from the picture of this question, we can find the input of op#21 connetcs to the output of op#20. So how about changing the name of output of op#23 to "fc7_stop" , and change the name of input of op#21 to "fc7_stop", sadly, it seems operator "StopGradient" is special, it becomes:
![image](https://user-images.githubusercontent.com/13573827/34868483-e86443ca-f7be-11e7-8f31-e5a1a2322a91.png)
StopGradient creates a new branch, here is my code:
```
train_model = model_helper.ModelHelper("train_net")

train_model.net = core.Net(predict_net_proto)

train_model.net.StopGradient("fc7","fc7_stop")

train_model.net.Proto().op[21].input[0] = "fc7_stop"
```

**Comment by harouwu** [comment]
Hi LMerCy,

Thanks for your question. First, if we have a net and we want to add stop gradient to fc7. We can do this:

fc7_w  ---\                                               fc_8_w     --- \
data    ----   fc_7_output  -- stopgradient--- fc_7_stop ---     fc_8_output
fc7_b   ---/                                               fc_8_b      ---/


Which means there would include

1. add StopGradient to fc_7_output
2. Change the fc_8 op input to fc_7_stop

This would change the proto. There is no wrapped function to do this. But you can check out how to change proto in python. It is not too hard. caffe2/python/core.py provides lot of example about it.

Yiming

**Comment by CriCL** [comment]
Hi @LMerCy , were you able to retrain fc7?

I'm close to finish the same goal, but having problems feeding "data" input

**Comment by LMerCy** [comment]
@CriCL  No, how do u insert stopgradient after fc7? What's the datatype of your data? U might have to pop the data init op from param_init_net first.Then you may create a new net to load the data and feed it to "data".
